### PR TITLE
fix: lazy issue in audit service

### PR DESCRIPTION
**Pull request by vietnguyen**
This is to rewrite/fix the audit service to avoid lazy exception during serialization.
The idea to to manually scan through each audit entity's properties and generate a serializable entry before sending it to jackson for json serialization. 
The current rule to filter out not-needed properties is using `owner` = true base on the schema of that property class.
Another added logic is to map all referenced BaseIdentifiableObject to only UID string. 

Number of hibernate queries generated during this process:
- 1 Select query to create the entity proxy. 
- n select queries to get the reference properties with `owner` = true.
If object doesn't have any lazy collections/properties, no select query will be triggered. 

`AuditIntegrationTest` covers some of the happy paths of this code

**Known issue**: This code doesn't take into account the Embedded object. If the property object extends `Embedded` then it will not be converted to UID string, in result when serializing  all lazy references of this property will throw LazyInitializationException if they are not loaded.  Will address this later in a separate PR

### Performance test result 
Before the fix
Screen Shot 2020-11-03 at 3 57 21 PM

After the fix
Screen Shot 2020-11-03 at 4 32 23 PM

### Debug using StopWatch, print time (ms) to serialize one object

Before the fix
```
class org.hisp.dhis.organisationunit.OrganisationUnit time : 34
class org.hisp.dhis.program.Program time : 102
class org.hisp.dhis.trackedentity.TrackedEntityInstance time : 28
```

After the fix
```
class org.hisp.dhis.trackedentity.TrackedEntityInstance time : 0
class org.hisp.dhis.trackedentityattributevalue.TrackedEntityAttributeValue time : 0
class org.hisp.dhis.program.Program time : 1
```

------------------------------
### Sample records

OrganisationUnit
` org.hisp.dhis.organisationunit.OrganisationUnit | O6uvpzGd5pu | {"uid": "O6uvpzGd5pu", "code": "OU_264", "attributeValues": []} | UPDATE    | {"lastUpdatedBy":"xE7jOejl9FI","parent":"ImspTQPwCqd","code":"OU_264","created":"2012-02-17T15:54:39.987","attributeValues":[],"description":"asdasdasd","uid":"O6uvpzGd5pu","lastUpdated":"2020-10-06T23:44:04.603","path":"/ImspTQPwCqd/O6uvpzGd5pu","translations":[],"name":"Bo","geometry":{},"shortName":"Bo","openingDate":"1990-02-01T00:00:00.000","user":"xE7jOejl9FI"}`

TrackedEntityInstance
`org.hisp.dhis.trackedentity.TrackedEntityInstance                     | F8TVn5sSZZ9 | {"uid": "F8TVn5sSZZ9", "deleted": false, "inactive": false, "attributeValues": [], "organisationUnit": "DiszpKrYNg8", "trackedEntityType": "nEenWmSyUEp"}
                                                                                 | READ      | {"created":"2020-10-05T22:28:35.920","lastUpdated":"2020-10-05T22:28:35.921","translations":[],"externalAccess":false,"userGroupAccesses":[],"userAccesses":[],
"favorites":[],"lastUpdatedBy":{"id":"xE7jOejl9FI","code":null,"name":"John Traore","displayName":"John Traore","username":"admin"},"deleted":false,"createdAtClient":"2020-10-05T22:28:35.920","lastUpdatedAtClient":"2020-10-05T22:28:35.920","relationshipI
tems":null,"programInstances":null,"programOwners":null,"organisationUnit":{"code":"OU_559","name":"Ngelehun CHC","created":"2012-02-17T15:54:39.987","lastUpdated":"2020-10-06T03:56:03.827","translations":[],"externalAccess":false,"user":{"id":"xE7jOejl9
FI","code":null,"name":"John Traore","displayName":"John Traore","username":"admin"},"userGroupAccesses":[],"userAccesses":[],"favorites":[],"lastUpdatedBy":{"id":"xE7jOejl9FI","code":null,"name":"John Traore","displayName":"John Traore","username":"admi
n"},"favorite":false,"displayName":"Ngelehun CHC","id":"DiszpKrYNg8","attributeValues":[]},"trackedEntityType":{"name":"Person","created":"2014-08-20T12:28:56.409","lastUpdated":"2018-02-22T09:05:13.160","translations":[],"externalAccess":false,"publicAc
cess":"rwrw----","user":{"id":"xE7jOejl9FI","code":null,"name":"John Traore","displayName":"John Traore","username":"admin"},"userGroupAccesses":null,"userAccesses":null,"favorites":[],"lastUpdatedBy":{"id":"xE7jOejl9FI","code":null,"name":"John Traore","displayName":"John Traore","username":"admin"},"description":"Person","trackedEntityTypeAttributes":null,"featureType":"NONE","minAttributesRequiredToSearch":1,"maxTeiCountToReturn":0,"allowAuditLog":false,"displayDescription":"Person","displayFormName":"Person","favorite":false,"displayName":"Person","id":"nEenWmSyUEp","attributeValues":[]},"inactive":false,"storedBy":"admin","favorite":false,"id":"F8TVn5sSZZ9","attributeValues":[],"trackedEntityAttributeValues":null}`

Program
` org.hisp.dhis.program.Program                   | IpHINAT79UW | {"uid": "IpHINAT79UW", "attributeValues": []}                   | UPDATE    | {"publicAccess":"rw------","description":"sadasd","completeEventsExpiryDays":0,"skipOffline":false,"ignoreOverd
ueEvents":false,"uid":"IpHINAT79UW","lastUpdated":"2020-10-06T23:44:36.050","notificationTemplates":["pwp8HUFajO1"],"categoryCombo":"bjDvmb4bfuf","translations":[],"programAttributes":["l2T72XzBCLd","XmRd2ZDjWhF","pWEEfUJUjej","ALw63dHrw1n"],"featureType
":"POINT","minAttributesRequiredToSearch":1,"organisationUnits":["z4silfLpw2G","NJolnlvYgLr","ZxuSbAmsLCn","wxMmC45UyNw","wjP03y8OY5k","Uo4cyJwAhTW","Eyqyhztf8G1","GyH8bjdOTsD","rpAgG9XCWhO","Zwnfm4rnzbZ","ZKL5hlVG6F6","oIgBLlEo6eH","Qwzs1iinAI7","Zp2Yi4j2AAH","K3k64jslIlL","si34vmovtgR","jGYT5U5qJP6","cWIiusmHULW","dczh6Jfd4no","zQ2pFkzGtIg","NLN0MvWv9tl","c9wCIfbcyVo","FsunWIQLXoF","uNEhNuBUr0i","DqfiI6NVnB1","cd3U2Tp0qR2","yJ1xkKha5oE","y5hLlID8ihI","GGDHb8xd8jc","VFF7f43dJv4","msH78gZ7Fe6","PMsF64R6OJX","IN2dOk0gY1G","DcmSvQd5N8c","PnMPARoMhWW","TGRCfJEnXJr","BJMWTGwuGiw","PWqwcBdRGIH","OI0BQUurVFS","O63vIA5MVn6","esMAQ4vs4kM","b09gf2vvZDb","ldXIdLNUNEn","e2WgqiasKnD","qELjt3LRkSD","k8ZPul89UDm","aSnKB1sWaz4","dtuiqEXYa7z","QZ5rmKrVleg","W7ekX3gi0ut","mTNOoGXuC39","ctN0WgIvfke","M9q1wOOsrXp","wjFsUXI1MlO","LV2b3vaLRl1","B9RxRfRUi2R","NqLYdlnK8sc","Pae8DR7VmcL","rebbn0ooFSO","cZxP4NE5O9z","u0SlCNJnK3K","TSyzvBiovKh","fdsRQbuuAuh","ubsjwFFBaJM","a04CZxe0PSe","BzEwqabuW19","XmfqaErvQ2T","YQ3csPLAlrn","CqARw68kXbB","foPGXhwhlqp","EQUwHqZOb5L","Gba5bTc8NIg","ewh5SKxcCAl","vyIl6s0lhKc","GM9ddjXIO5b","OqBiNJjKQAu","sDTodaygv5u","X9zzzyPZViR","nYiOoF2nXIr","suFG8zx4bU3","pmzk0ho80aA","QMnoFLTLpkY","Tc3zugEWdTm","oV9P0VvL9Jh","HC2NlwpoXfb","UJ80rknbJtm","OZ1olxsTyNa","JrSIoCOdTH2","uYTq3TEO2a9","Ahh47q8AkId","BXJnMD2eJAx","InQWjSe6k2f","sYJCxNdKHxR","CgunjDKbM45","VdXuxcNkiad"],"displayFrontPageList":true,"enrollmentDateLabel":"Date of enrollment","lastUpdatedBy":"xE7jOejl9FI","programType":"WITH_REGISTRATION","onlyEnrollOnce":true,"userGroupAccesses":[{"access":"r-rw----","userGroupUid":"M1Qre0247G3","displayName":"_PROGRAM_Antenatal care program","id":"M1Qre0247G3"},{"access":"r-rw----","userGroupUid":"H9XnHoWRKCg","displayName":"_PROGRAM_Child Health Tracker","id":"H9XnHoWRKCg"},{"access":"r-rw----","userGroupUid":"vRoAruMnNpB","displayName":"_PROGRAM_MNCH / PNC (Adult Woman) program","id":"vRoAruMnNpB"},{"access":"r-rw----","userGroupUid":"Kk12LkEWtXp","displayName":"_PROGRAM_TB program","id":"Kk12LkEWtXp"},{"access":"r-rw----","userGroupUid":"z1gNAf2zUxZ","displayName":"_PROGRAM_WHO MCH program","id":"z1gNAf2zUxZ"},{"access":"r-rw----","userGroupUid":"gXpmQO6eEOo","displayName":"_PROGRAM_Superuser","id":"gXpmQO6eEOo"},{"access":"r-rw----","userGroupUid":"pBnkuih0c1K","displayName":"_PROGRAM_System administrator (ALL)","id":"pBnkuih0c1K"},{"access":"r-rw----","userGroupUid":"NTC8GjJ7p8P","displayName":"_PROGRAM_Inpatient program","id":"NTC8GjJ7p8P"}],"accessLevel":"OPEN","created":"2013-03-04T11:41:07.494","programSections":[],"attributeValues":[],"version":37,"maxTeiCountToReturn":0,"selectIncidentDatesInFuture":false,"incidentDateLabel":"Date of birth","programStages":["A03MvHHogjR","ZzYYXq4fJie"],"trackedEntityType":"nEenWmSyUEp","displayIncidentDate":true,"userAccesses":[],"name":"Child Programme","selectEnrollmentDatesInFuture":false,"style":{"color":"#fbc02d","icon":"child_program_positive"},"expiryDays":0,"shortName":"Child Programme","user":"xE7jOejl9FI","useFirstStageDuringRegistration":true}`

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/AbstractHibernateListener.java [hunk 1]
```diff
@@ -102,63 +109,126 @@ public String getCreatedBy()
...
+    protected Object createAuditEntry( PostUpdateEvent postUpdateEvent )
     {
-        Session session = factory.getCurrentSession();
+        return createAuditEntry( postUpdateEvent.getEntity(), postUpdateEvent.getState(), postUpdateEvent.getSession(), postUpdateEvent.getId(), postUpdateEvent.getPersister() );
+    }
+
+    /**
+     * Create Audit entry for insert event
+     */
+    protected Object createAuditEntry( PostInsertEvent postInsertEvent )
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/AbstractHibernateListener.java:122]: I would move "down" these two methods (`createAuditEntry`) to the sub-classes (`PostInsertAuditListener` and `PostUpdateAuditListener`)

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostDeleteAuditListener.java [hunk 2]
```diff
@@ -28,20 +28,32 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
-import org.hibernate.SessionFactory;
+import org.hibernate.Hibernate;
+import org.hibernate.collection.spi.PersistentCollection;
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostDeleteAuditListener.java:34]: Please fix unused comments, no star import

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostInsertAuditListener.java [hunk 3]
```diff
@@ -28,20 +28,32 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
-import org.hibernate.SessionFactory;
+import org.hibernate.Hibernate;
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostInsertAuditListener.java:33]: Please fix unused comments, no star import

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostUpdateAuditListener.java [hunk 4]
```diff
@@ -29,19 +29,29 @@
  */
 
 import lombok.extern.slf4j.Slf4j;
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostUpdateAuditListener.java:31]: Please fix unused comments, no star import

File: dhis-2/dhis-support/dhis-support-commons/src/main/java/org/hisp/dhis/commons/config/JacksonObjectMapperConfig.java [hunk 5]
```diff
@@ -28,6 +28,17 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+import java.util.Date;
+
+import org.hibernate.SessionFactory;
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-commons/src/main/java/org/hisp/dhis/commons/config/JacksonObjectMapperConfig.java:33]: Please fix unused comments

**Review by luciano-fiandesio — CHANGES_REQUESTED** [review]
hi @vietnguyen, sorry for the late review.
In general, I think your approach looks very good. 
I'm only worried about how this approach will perform under high concurrency.
It would be probably good to stress test the system to check  if there is any significant performance degradation. Let's not forget that this part of the code is very "hot", if auditing is enabled. We noticed in 2.35.0, that when audit is enabled, the performance of adding a new Event are 50% slower!

**Review by netroms — APPROVED** [review]

**Comment by netroms** [comment]
This looks very good, only thing I miss is a couple of unit tests (happy path)

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/AuditableEntity.java [hunk 6]
```diff
@@ -39,4 +39,16 @@
 public class AuditableEntity
 {
     private Object entity;
+
+    private Object serializableObject;
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/AuditableEntity.java]: fixed

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/AuditableEntity.java [hunk 7]
```diff
@@ -39,4 +39,16 @@
 public class AuditableEntity
 {
     private Object entity;
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/AuditableEntity.java]: Replaced Object entity with Class entityClass. 
We need this Class metadata for annotation scanning

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/AbstractHibernateListener.java [hunk 8]
```diff
@@ -28,23 +28,31 @@
...
 
-import org.apache.commons.logging.Log;
-import org.apache.commons.logging.LogFactory;
-import org.hibernate.LazyInitializationException;
+import com.sun.tools.rngom.parse.host.Base;
+import org.hibernate.Hibernate;
 import org.hibernate.Session;
-import org.hibernate.SessionFactory;
-import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.event.spi.*;
```
[the remark this answers is not in the document]
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/AbstractHibernateListener.java]: fixed

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/AbstractHibernateListener.java (hunk 1 above)
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/AbstractHibernateListener.java]: fixed

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostDeleteAuditListener.java (hunk 2 above)
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostDeleteAuditListener.java]: fixed

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostInsertAuditListener.java (hunk 3 above)
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostInsertAuditListener.java]: fixed

File: dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostUpdateAuditListener.java (hunk 4 above)
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-artemis/src/main/java/org/hisp/dhis/artemis/audit/listener/PostUpdateAuditListener.java:31]: fixed

File: dhis-2/dhis-support/dhis-support-commons/src/main/java/org/hisp/dhis/commons/config/JacksonObjectMapperConfig.java (hunk 5 above)
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-commons/src/main/java/org/hisp/dhis/commons/config/JacksonObjectMapperConfig.java:33]: fixed

**Comment by vietnguyen** [comment]
> hi @vietnguyen, sorry for the late review.
> In general, I think your approach looks very good.
> I'm only worried about how this approach will perform under high concurrency.
> It would be probably good to stress test the system to check if there is any significant performance degradation. Let's not forget that this part of the code is very "hot", if auditing is enabled. We noticed in 2.35.0, that when audit is enabled, the performance of adding a new Event are 50% slower!

@luciano-fiandesio  
Fixed all of your feedbacks.
I'm positive that this fix will increase the performance of the old code. 
Will do stress test on this branch and update the result.

**Comment by vietnguyen** [comment]
> This looks very good, only thing I miss is a couple of unit tests (happy path)

There is a `AuditIntegrationTest` that covers some of the happy path of this audit service

**Comment by vietnguyen** [comment]
@luciano-fiandesio 
Below is the test result, please check

Locust 
Before the fix 
Screen Shot 2020-11-03 at 3 57 21 PM


After the fix

Screen Shot 2020-11-03 at 4 32 23 PM



Code debug using stopWatch
Before the fix
```
class org.hisp.dhis.organisationunit.OrganisationUnit time : 34
class org.hisp.dhis.program.Program time : 102
class org.hisp.dhis.trackedentity.TrackedEntityInstance time : 28
```

After the fix
```
class org.hisp.dhis.trackedentity.TrackedEntityInstance time : 0
class org.hisp.dhis.trackedentityattributevalue.TrackedEntityAttributeValue time : 0
class org.hisp.dhis.program.Program time : 1
class org.hisp.dhis.organisationunit.OrganisationUnit time : 1
```

**Review by netroms — APPROVED** [review]

**Review by mortenoh — APPROVED** [review]

**Pull request merged**